### ISSUE TITLE
Neither Nat not Nat in datatype doesn't work

### ISSUE THREAD

**Issue by uzytkownik** [goldfirere/singletons#478]
```haskell
$(singletons [d|
        data Foo = Foo Nat
        data Bar = Bar Natural
    |])
```

It produces following code:

```haskell
    data Foo_aCRh = Foo_aCRi Nat
    data Bar_aCRj = Bar_aCRk Natural
    type FooSym1 (t6989586621679159182 :: Nat) =
        Foo_aCRi t6989586621679159182
    instance SuppressUnusedWarnings FooSym0 where
      suppressUnusedWarnings = snd (((,) FooSym0KindInference) ())
    data FooSym0 :: (~>) Nat Foo_aCRh
      where
        FooSym0KindInference :: forall t6989586621679159182
                                       arg_aCRx. SameKind (Apply FooSym0 arg_aCRx) (FooSym1 arg_aCRx) =>
                                FooSym0 t6989586621679159182
    type instance Apply FooSym0 t6989586621679159182 = Foo_aCRi t6989586621679159182
    type BarSym1 (t6989586621679159184 :: Natural) =
        Bar_aCRk t6989586621679159184
    instance SuppressUnusedWarnings BarSym0 where
      suppressUnusedWarnings = snd (((,) BarSym0KindInference) ())
    data BarSym0 :: (~>) Natural Bar_aCRj
      where
        BarSym0KindInference :: forall t6989586621679159184
                                       arg_aCRz. SameKind (Apply BarSym0 arg_aCRz) (BarSym1 arg_aCRz) =>
                                BarSym0 t6989586621679159184
    type instance Apply BarSym0 t6989586621679159184 = Bar_aCRk t6989586621679159184
    data SFoo :: Foo_aCRh -> ghc-prim-0.5.3:GHC.Types.Type
      where
        SFoo :: forall (n_aCRA :: Nat).
                (Sing (n_aCRA :: Nat)) -> SFoo (Foo_aCRi n_aCRA)
    type instance Sing @Foo_aCRh = SFoo
    instance SingKind Foo_aCRh where
      type Demote Foo_aCRh = Foo_aCRh
      fromSing (SFoo b_aCRC) = Foo_aCRi (fromSing b_aCRC)
      toSing (Foo_aCRi (b_aCRE :: Demote Nat))
        = case toSing b_aCRE :: SomeSing Nat of {
            SomeSing c_aCRF -> SomeSing (SFoo c_aCRF) }
    data SBar :: Bar_aCRj -> ghc-prim-0.5.3:GHC.Types.Type
      where
        SBar :: forall (n_aCRH :: Natural).
                (Sing (n_aCRH :: Natural)) -> SBar (Bar_aCRk n_aCRH)
    type instance Sing @Bar_aCRj = SBar
    instance SingKind Bar_aCRj where
      type Demote Bar_aCRj = Bar_aCRj
      fromSing (SBar b_aCRJ) = Bar_aCRk (fromSing b_aCRJ)
      toSing (Bar_aCRk (b_aCRL :: Demote Natural))
        = case toSing b_aCRL :: SomeSing Natural of {
            SomeSing c_aCRM -> SomeSing (SBar c_aCRM) }
    instance SingI n_aCRA => SingI (Foo_aCRi (n_aCRA :: Nat)) where
      sing = SFoo sing
    instance SingI (FooSym0 :: (~>) Nat Foo_aCRh) where
      sing = (singFun1 @FooSym0) SFoo
    instance SingI n_aCRH => SingI (Bar_aCRk (n_aCRH :: Natural)) where
      sing = SBar sing
    instance SingI (BarSym0 :: (~>) Natural Bar_aCRj) where
      sing = (singFun1 @BarSym0) SBar
```

However both produce errors:

```
<interactive>:25:3: error:
    • Could not deduce: Demote Natural ~ Natural
      from the context: a ~ 'Bar n
        bound by a pattern with constructor:
                   SBar :: forall (n :: Natural). Sing n -> SBar ('Bar n),
                 in an equation for ‘fromSing’
        at <interactive>:(25,3)-(28,6)
    • In the first argument of ‘Bar’, namely ‘(fromSing b_aAHx)’
      In the expression: Bar (fromSing b_aAHx)
      In an equation for ‘fromSing’:
          fromSing (SBar b_aAHx) = Bar (fromSing b_aAHx)

<interactive>:25:3: error:
    • Couldn't match expected type ‘Natural’
                  with actual type ‘Demote Natural’
    • When checking that the pattern signature: Demote Natural
        fits the type of its context: Natural
      In the pattern: b_aAHz :: Demote Natural
      In the pattern: Bar (b_aAHz :: Demote Natural)

<interactive>:25:3: error:
    • Couldn't match type ‘Natural’ with ‘Nat’
      Expected type: Nat
        Actual type: Demote Nat
    • In the first argument of ‘Foo’, namely ‘(fromSing b_aAHq)’
      In the expression: Foo (fromSing b_aAHq)
      In an equation for ‘fromSing’:
          fromSing (SFoo b_aAHq) = Foo (fromSing b_aAHq)

<interactive>:25:3: error:
    • Couldn't match type ‘Nat’ with ‘Natural’
      Expected type: Nat
        Actual type: Demote Nat
    • When checking that the pattern signature: Demote Nat
        fits the type of its context: Nat
      In the pattern: b_aAHs :: Demote Nat
      In the pattern: Foo (b_aAHs :: Demote Nat)
```

**Comment by RyanGlScott** [comment]
This is a well-known problem with using `Nat` in the context of `singletons`, since `Natural` and `Nat` are distinct data types.  The recommended workaround is to define separate versions of your data type, where one version uses `Natural` and the other uses `Nat`, and convert from one to the other during promotion. I've documented how to do this in a semi-automated fashion in [this section of the `README`](https://github.com/goldfirere/singletons/tree/0261bafe8fe96944d1ee5bd44a6beaf12dc87e3e#arrows-nat-symbol-and-literals). Does this answer your question?

Note that a future version of GHC will [merge these data types](https://github.com/goldfirere/singletons/issues/477), which will remove the need for this workaround. But that is still a ways off.

**Comment by RyanGlScott** [comment]
My apologies, I now realize that the section of the `README` that I linked to in https://github.com/goldfirere/singletons/issues/478#issuecomment-736405895 relies on new `singletons` Template Haskell features that haven't been released to Hackage yet. (Hopefully soon, but not yet.) If you don't want to wait for the next `singletons` Hackage release, you will likely have to write promoted and singled code involving `Nat`/`Natural` by hand. Sorry!

**Comment by RyanGlScott** [comment]
I've released `singletons-{,th-base}-3.0` to Hackage, which implements the workaround described in https://github.com/goldfirere/singletons/issues/478#issuecomment-736405895. I'll close this as a result, but feel free to re-open this if this does not address your needs.
